src/param_env.md

```diff
@@ -1,69 +1,159 @@
-# Parameter Environment
+# The `ParamEnv` type
 
-When working with associated and/or generic items (types, constants,
-functions/methods) it is often relevant to have more information about the
-`Self` or generic parameters. Trait bounds and similar information is encoded in
-the [`ParamEnv`][pe]. Often this is not enough information to obtain things like the
-type's `Layout`, but you can do all kinds of other checks on it (e.g. whether a
-type implements `Copy`) or you can evaluate an associated constant whose value
-does not depend on anything from the parameter environment.
+## What is it
 
-[pe]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.ParamEnv.html
+The type system relies on information in the environment in order for it to function correctly. This information is stored in the [`ParamEnv`][pe] type and it is important to use the correct `ParamEnv` when interacting with the type system.
 
-For example if you have a function
+The information represented by `ParamEnv` is a list of in scope where clauses, and a [`Reveal`][reveal]. A `ParamEnv` typically corresponds to a specific item's environment however it can also be created with arbitrary data that is not derived from a specific item. In most cases `ParamEnv`s are initially created via the [`param_env` query][query] which returns a `ParamEnv` derived from the provided item's where clauses.
 
+If we have a function such as:
 ```rust
-fn foo<T: Copy>(t: T) { ... }
+// `foo` would have a `ParamEnv` of:
+// `[T: Sized, T: Trait, <T as Trait>::Assoc: Clone]`
+fn foo<T: Trait>()
+where
+    <T as Trait>::Assoc: Clone,
+{}
 ```
+If we were interacting with the type system from inside of `foo` we would use this `ParamEnv` everywhere that we interact with the type system. This would allow things such as normalization (TODO: write a chapter about normalization and link it), evaluating generic constants, and proving where clauses/goals, to rely on `T` being sized, implementing `Trait`, etc.
```

Comment:
```suggestion
If we were interacting with the type system (for example, type-checking or implementing a lint) from inside of `foo` we would use this `ParamEnv` everywhere that we interact with the type system. This would allow things such as normalization (TODO: write a chapter about normalization and link it), evaluating generic constants, and proving where clauses/goals, to rely on `T` being sized, implementing `Trait`, etc.
```